### PR TITLE
Add out-of-order sample support to the TSDB

### PR DESCRIPTION
**Pull request by codesome**
This implementation is based on this design doc:
https://docs.google.com/document/d/1Kppm7qL9C-BJB1j6yb6-9ObG3AbdZnFUBYPNNWwDBYM/edit?usp=sharing

This commit adds support to accept out-of-order ("OOO") sample into the TSDB
up to a configurable time allowance. If OOO is enabled, overlapping querying
are automatically enabled.

### CODE REVIEW

**Comment by codesome** [comment]
This replaces https://github.com/grafana/mimir-prometheus/pull/268. Somehow having a single commit while manually adding co-authors removes the CLA requirement from the co-authors. (See 268, CLA fails because Dieter hasn't signed, but it does not fail here).

**Comment by codesome** [comment]
We can consider merging this after https://github.com/grafana/mimir-prometheus/pull/270/ is merged and synced in this PR.

**Comment by jesusvazquez** [comment]
Lets get the OOOAllowance and the OOO WBL dir rename into this PR.

File: tsdb/head_read.go [hunk 1]
```diff
@@ -358,6 +367,259 @@ func (s *memSeries) chunk(id chunks.HeadChunkID, cdm chunkDiskMapper) (chunk *me
...
+
+func (o mergedOOOChunks) NumSamples() int {
+	samples := 0
+	for _, c := range o.chunks {
+		samples += c.Chunk.NumSamples()
+	}
+	return samples
+}
+
+// Compact TODO(jesus.vazquez) Clarify why this method wont be called
```
Comment (jesusvazquez) [inline tsdb/head_read.go:544]: Maybe we can remove this

File: tsdb/index/postings.go [hunk 2]
```diff
@@ -49,8 +49,12 @@ var ensureOrderBatchPool = sync.Pool{
 // EnsureOrder() must be called once before any reads are done. This allows for quick
 // unordered batch fills on startup.
 type MemPostings struct {
-	mtx     sync.RWMutex
-	m       map[string]map[string][]storage.SeriesRef
+	mtx sync.RWMutex
+	m   map[string]map[string][]storage.SeriesRef
+	// TODO We should have a mapping of which series have out of order metrics
```
Comment (jesusvazquez) [inline tsdb/index/postings.go:54]: I don't think we need this anymore?

File: tsdb/index/postings.go [hunk 3]
```diff
@@ -216,6 +220,8 @@ func (p *MemPostings) Get(name, value string) Postings {
 	return newListPostings(lp...)
 }
 
+// TODO GetOOOPostings( returns newOOOListPostings )
```
Comment (jesusvazquez) [inline tsdb/index/postings.go:223]: Remove this too

File: tsdb/ooo_head.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+		Stats: BlockStats{
+			NumSeries: oh.head.NumSeries(),
+		},
+	}
+}
+
+// Size returns 0 because the space taken by the out of order samples is taken
+// into account by the RangeHead size.
+func (oh *OOORangeHead) Size() int64 {
+	// TODO(jesus.vazquez) Find what's the appropriate value here
```
Comment (jesusvazquez) [inline tsdb/ooo_head.go:59]: We never solved this, i think its also not called. WDYT @codesome

**Review by jesusvazquez — COMMENTED** [review]
I did take a look at some of the TODO comments that we can get rid of. Can you take a look at the comments @codesome?

File: tsdb/ooo_head.go (hunk 4 above)
Comment (codesome) [reply tsdb/ooo_head.go:59]: Oh, that makes me realise that we are not including the size of WBL anywhere. I will fix that.

**Comment by codesome** [comment]
Opened https://github.com/grafana/mimir-prometheus/pull/273 for TODOs

**Review by jesusvazquez — APPROVED** [review]
Preemptively approving knowing that you will rebase this PR with the changes from https://github.com/grafana/mimir-prometheus/pull/273

Let's goooo :tada: :tada: :tada: :tada:

**Comment by nwpuCfy** [comment]
> This PR changes `ChunkReader` interface in a way that's incompatible with usage in Thanos and Mimir. We could fix Mimir, but unless you plan to vendor `mimir-prometheus` in Thanos, it needs to be reverted :(

Did this mean TSDB that supports out of order data is not compatible with thanos? So I can't use it in cortex v1.11 that supports obs storage?

**Comment by codesome** [comment]
@nwpuCfy that comment was about code interface and not the data format. Thanos has switched to the new interface I think. And they use prometheus/prometheus, and not grafana/mimir-prometheus.

**Comment by jesusvazquez** [comment]
@nwpuCfy As @codesome pointed out, thanos is now vendoring prometheus/prometheus were we ported the changes you see in this PR so there should not be any problems :+1:  Use prometheus/prometheus as your reference for the official state of out-of-order support.

**Comment by nwpuCfy** [comment]
> @nwpuCfy As @codesome pointed out, thanos is now vendoring prometheus/prometheus were we ported the changes you see in this PR so there should not be any problems 👍 Use prometheus/prometheus as your reference for the official state of out-of-order support.
ok, thanks for reply

**Comment by nwpuCfy** [comment]
> @nwpuCfy that comment was about code interface and not the data format. Thanos has switched to the new interface I think. And they use prometheus/prometheus, and not grafana/mimir-prometheus.

thanks for reply

**Pull request merged**